llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int.ll

```diff
@@ -1155,8 +1155,8 @@ define void @mulhu_v8i16(ptr %x) {
 ; CHECK-NEXT:    vle16.v v8, (a0)
 ; CHECK-NEXT:    vmv.v.i v9, 0
 ; CHECK-NEXT:    lui a1, 1048568
-; CHECK-NEXT:    vsetvli zero, zero, e16, m1, tu, ma
 ; CHECK-NEXT:    vmv.v.i v10, 0
```

Comment:
This test change might be worth a second look.  Can't tell in isolation if this is correct or not, but the core point is that I don't understand why this one is happening.

Reply:
The difference in this test case is caused by the twoaddr pass splitting the `PseudoVMV_S_X` into two pseudo-instructions. However, the first pseudo-instruction lacks the SEW for the vsetvli pass to recognize and emit the appropriate vsetvl instruction.

After the two addr pass

```
%4:vr = PseudoVMV_S_X %2:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype
-> 
%4:vr = COPY %2:vr
%4:vr = PseudoVMV_S_X %4:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype
```

After Post-RA pseudo instruction expansion

```
renamable $v10 = COPY renamable $v9
-> 
$v10 = PseudoVMV_V_I_M1 undef $v10(tied-def 0), 0, 8, 4, 0, implicit $vl, implicit $vtype
```

---

vsetvl insert before twoaddr

```

%4:vr = PseudoVMV_S_X %2:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype

-> vsetvl insertion 

dead $x0 = PseudoVSETVLIX0 killed $x0, 136, implicit-def $vl, implicit-def $vtype, implicit $vl
%4:vr = PseudoVMV_S_X %2:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype

-> two-addr

dead $x0 = PseudoVSETVLIX0 killed $x0, 136, implicit-def $vl, implicit-def $vtype, implicit $vl
%4:vr = COPY %2:vr
%4:vr = PseudoVMV_S_X %4:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype

-> post-ra pesudo expand

dead $x0 = PseudoVSETVLIX0 killed $x0, 136, implicit-def $vl, implicit-def $vtype, implicit $vl
$v10 = PseudoVMV_V_I_M1 undef $v10(tied-def 0), 0, 8, 4, 0, implicit $vl, implicit $vtype
renamable $v10 = PseudoVMV_S_X killed renamable $v10(tied-def 0), killed renamable $x11, 8, 4, implicit $vl, implicit $vtype
```

vsetvl insert after twoaddr

```
%4:vr = PseudoVMV_S_X %2:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype

-> two-addr

%4:vr = COPY %2:vr
%4:vr = PseudoVMV_S_X %4:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype

->vsetvl insertion 

%4:vr = COPY %2:vr
dead $x0 = PseudoVSETVLIX0 killed $x0, 136, implicit-def $vl, implicit-def $vtype, implicit $vl
%4:vr = PseudoVMV_S_X %4:vr(tied-def 0), killed %3:gpr, 8, 4, implicit $vl, implicit $vtype

-> post-ra pesudo expand

$v10 = PseudoVMV_V_I_M1 undef $v10(tied-def 0), 0, 8, 4, 0, implicit $vl, implicit $vtype
dead $x0 = PseudoVSETVLIX0 killed $x0, 136, implicit-def $vl, implicit-def $vtype, implicit $vl
renamable $v10 = PseudoVMV_S_X killed renamable $v10(tied-def 0), killed renamable $x11, 8, 4, implicit $vl, implicit $vtype
```

Reply:
How come in both cases we're inserting a PseudoVMV_V_I_M1 after vsetvl insertion? Wouldn't it need to know about VL/VTYPE and potentially insert a vsetvli? I thought it would have only emitted vmv1r.v etc.

Reply:
Maybe it assume VL/VTYPE dependant doesn't break?

> NOTE:
> post-ra pseudo pass invoke `TII->lowerCOPY -> RISCVInstrInfo::copyPhysReg -> RISCVInstrInfo::copyPhysRegVector` to > emit the `PseudoVMV_V_I_M1` base on `COPY`.

Reply:
> Maybe it assume VL/VTYPE dependant doesn't break?
> 
> > NOTE:
> > post-ra pseudo pass invoke `TII->lowerCOPY -> RISCVInstrInfo::copyPhysReg -> RISCVInstrInfo::copyPhysRegVector` to > emit the `PseudoVMV_V_I_M1` base on `COPY`.

No. Machine scheduler could move the COPY before vsetvl. VL/VTYPE dependent doesn't constrain COPY at all. The current scheduled order base on input order. 

```
SU(3):   %6:vr = PseudoVMV_V_I_M1 undef %6:vr(tied-def 0), 0, 8, 4, 0, implicit $vl, implicit $vtype

SU(7) %4:vr = PseudoVMV_S_X %4:vr(tied-def 0), %3:gpr, 8, 4, implicit $vl, implicit $vtype
SU(11) early-clobber %6:vr = PseudoVSLIDEUP_VI_M1 %6:vr(tied-def 0), %5:vr, 6, 7, 4, 0, implicit $vl, implicit $vtype

SU(6):   %4:vr = COPY %6:vr
  Predecessors:
    SU(3): Data Latency=1 Reg=%6
  Successors:
    SU(7): Data Latency=0 Reg=%4
    SU(7): Out  Latency=1
    SU(11): Anti Latency=0
```

Reply:
Ah it actually looks for the last vsetvli to see if the SEW/VTYPE matches in `isConvertibleToVMV_V_V`. Thanks for the explanation, it makes sense to me now. The `vmv.v.i v10, 0` is a copy of the `vmv.v.i v9, 0` above. And because the `vmv.v.i v9, 0` has a TA policy, RISCVInstrInfo::copyPhysRegVector can insert the `vmv.v.i 10, 0` in either a TU or TA policy.

Reply:
Thank you both for the analysis here.  Luke, your last comment explains this sufficiently for me.